packages/services/src/contents/index.ts

```diff
@@ -1571,3 +1578,233 @@ namespace Private {
+      // The notebook type cannot take a format option.
+      if (options.type === 'notebook') {
+        delete options['format'];
+      }
+      const content = options.content ? '1' : '0';
+      const hash = options.hash ? '1' : '0';
+      const params: PartialJSONObject = { ...options, content, hash };
+      url += URLExt.objectToQueryString(params);
+    }
+
+    const settings = this._options.serverSettings;
+    const response = await ServerConnection.makeRequest(url, {}, settings);
+    if (response.status !== 200) {
+      const err = await ServerConnection.ResponseError.create(response);
+      throw err;
+    }
+    const data = await response.json();
+    validate.validateContentsModel(data);
+    return data;
+  }
+
+  /**
+   * Save a file.
+   *
+   * @param localPath - The desired file path.
+   *
+   * @param options - Optional overrides to the model.
+   *
+   * @returns A promise which resolves with the file content model when the
+   *   file is saved.
+   *
+   * #### Notes
+   * Ensure that `model.content` is populated for the file.
+   *
+   * Uses the [Jupyter Server API](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter-server/jupyter_server/main/jupyter_server/services/api/api.yaml#!/contents) and validates the response model.
+   */
+  async save(
+    localPath: string,
+    options: Partial<Contents.IModel> = {}
+  ): Promise<Contents.IModel> {
+    const settings = this._options.serverSettings;
+    const url = this._getUrl(localPath);
+    const init = {
+      method: 'PUT',
+      body: JSON.stringify(options)
+    };
+    const response = await ServerConnection.makeRequest(url, init, settings);
+    // will return 200 for an existing file and 201 for a new file
+    if (response.status !== 200 && response.status !== 201) {
+      const err = await ServerConnection.ResponseError.create(response);
+      throw err;
+    }
+    const data = await response.json();
+    validate.validateContentsModel(data);
+    this.fileChanged.emit({
+      type: 'save',
+      oldValue: null,
+      newValue: data
+    });
+    return data;
+  }
+
+  /**
+   * A signal emitted when a file operation takes place.
+   */
+  get fileChanged(): Signal<this, Contents.IChangedArgs> {
+    return this._options.fileChanged as unknown as Signal<
+      this,
+      Contents.IChangedArgs
+    >;
+  }
+
+  /**
+   * Get a REST url for a file given a path.
+   */
+  private _getUrl(...args: string[]): string {
+    const parts = args.map(path => URLExt.encodeParts(path));
+    const baseUrl = this._options.serverSettings.baseUrl;
+    return URLExt.join(baseUrl, this._options.apiEndpoint, ...parts);
+  }
+
+  private _options: RestContentProvider.IOptions;
+}
+
+namespace RestContentProvider {
+  export interface IOptions {
+    apiEndpoint: string;
+    serverSettings: ServerConnection.ISettings;
+    fileChanged: ISignal<Contents.IDrive, Contents.IChangedArgs>;
+  }
+}
+
+/**
+ * Registry of content providers.
```

Comment:
Should we add tag the APIs as experimental for now?

```suggestion
 * Registry of content providers.
 * @experimental
```